### PR TITLE
Fix bugs in module specifier generation with `paths`/`typesVersions`

### LINKED ISSUE #49034: Incorrect auto-import module specifier generated from typesVersions
# Bug Report

The `typesVersions` from unified@9 creates an auto-import suggestion that doesn’t resolve:

```ts
// @Filename: /node_modules/unified/package.json
//// {
////   "name": "unified",
////   "types": "types/ts3.4/index.d.ts",
////   "typesVersions": {
////     ">=4.0": {
////       "types/ts3.4/*": [
////         "types/ts4.0/*"
////       ]
////     }
////   }
//// }
```

image

### LINKED ISSUE #49290: [Node16/NodeNext] Automatic imports that are mapped in `imports` in `package.json` get extension appended to them, which causes a compile error
# Bug Report

### 🔎 Search Terms

import mappings, nodenext, node16, import extensions placed wrong

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ

### ⏯ ~~Playground~~ Repository Link

Repository where the issue can be seen: https://github.com/vladfrangu/typescript-import-mapping-issue

### 💻 Code

> package.json
```json
{
	"type": "module",
	"scripts": {
		"build": "tsc"
	},
	"devDependencies": {
		"@sapphire/ts-config": "^3.3.4",
		"@types/node": "^17.0.35",
		"typescript": "^4.7.2"
	},
	"imports": {
		"#internals/*": "./dist/internals/*.js"
	}
}
```

> tsconfig.json
```json
{
	"extends": "@sapphire/ts-config",
	"compilerOptions": {
		"target": "ESNext",
		"module": "Node16",
		"moduleResolution": "Node16",
		"rootDir": "./src",
		"outDir": "./dist",
		"paths": {
			"#internals/*": ["./src/internals/*.ts"]
		}
	},
	"include": ["src", "src/**/*.d.ts"]
}
```

> src/internals/example.ts
```ts
export function helloWorld() {
	console.log('Hello world!');
}
```

> src/index.ts
```ts
// Attempt to import this method with your editor with the "typescript.preferences.importModuleSpecifierEnding" set to "auto"
// Then try the same autoimport with the "typescript.preferences.importModuleSpecifierEnding" set to "index"
helloWorld();
```

### 🙁 Actual behavior

Given a `package.json` with import mappings, attempting to import a file defined in the mapping while `typescript.preferences.importModuleSpecifierEnding` is set to "auto" will cause the extension to be appended to the import, which in turn immediately causes TypeScript to report a `Cannot find module '#internals/example.js' or its corresponding type declarations.`.

Setting `typescript.preferences.importModuleSpecifierEnding` to "index" *works* but it will have side effects for relative imports, which is not ideal.

### 🙂 Expected behavior

TypeScript should only append the extension to imports that are relative, and not from imports seen from import mapping/paths. (maybe this also ties into #48120 but I am not sure if they're separate issues)

### PR DESCRIPTION
**Pull request by andrewbranch**
This was much more difficult to do correctly than I expected, but it fixes more than what I set out to fix.

Fixes #49034
Fixes #49290

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot perf test this

File: src/compiler/moduleSpecifiers.ts [hunk 1]
```diff
@@ -677,10 +749,10 @@ namespace ts.moduleSpecifiers {
 
         // Simplify the full file path to something that can be resolved by Node.
 
+        const preferences = getPreferences(host, userPreferences, options, importingSourceFile);
```
Comment (jakebailey) [inline src/compiler/moduleSpecifiers.ts:752]: Is this used outside the if block, or is it so cheap that it just doesn't matter where it is?

**Review by jakebailey — APPROVED** [review]
I think Wes is out (and you mentioned this in Please Review); I'm not an expert but it seems good to me and I appreciate the essay explaining what the deal is.

File: src/compiler/moduleSpecifiers.ts (hunk 1 above)
Comment (andrewbranch) [reply src/compiler/moduleSpecifiers.ts:752]: I assume I had a reason to move it; maybe refactored away in later edits, but it is trivially cheap.

File: src/compiler/moduleSpecifiers.ts [hunk 2]
```diff
@@ -760,19 +830,31 @@ namespace ts.moduleSpecifiers {
...
+                    // The 'main' file is also subject to mapping through typesVersions, and we couldn't come up with a path
+                    // explicitly through typesVersions, so if it matches a key in typesVersions now, it's not reachable.
+                    // (The only way this can happen is if some file in a package that's not resolvable from outside the
+                    // package got pulled into the program anyway, e.g. transitively through a file that *is* reachable. It
+                    // happens very easily in fourslash tests though, since every test file listed gets included. See
+                    // importNameCodeFix_typesVersions.ts for an example.)
                     const mainExportFile = toPath(mainFileRelative, packageRootPath, getCanonicalFileName);
                     if (removeFileExtension(mainExportFile) === removeFileExtension(getCanonicalFileName(moduleFileToTry))) {
+                        // ^ An arbitrary removal of file extension for this comparison is almost certainly wrong
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/moduleSpecifiers.ts:857]: More of a hint for a future developer investigating bugs in this area—a TODO only if prompted by a user report.

**Comment by andrewbranch** [comment]
Thanks for the review! I forgot this was the one with the big table. Writing that out actually helped my clarify my mental model of module specifier generation a lot, which was a topic I thought I already understood well.

**Pull request merged**